packages/contract/src/DaimoUSDCSwapper.sol

```diff
@@ -0,0 +1,302 @@
+import "@uniswap/v3-periphery/contracts/libraries/TransferHelper.sol";
+import "@uniswap/v3-periphery/contracts/interfaces/ISwapRouter.sol";
+import "@uniswap/v3-periphery/contracts/libraries/Path.sol";
+import "openzeppelin-contracts/contracts/token/ERC20/IERC20.sol";
+import "./IDaimoSwapper.sol";
+
+// Fully on-chain swap executor that quotes stablecoin swaps using Uniswap V3
+// TWAP/TWALs. Seehttps://uniswap.org/whitepaper-v3.pdf for more on TWAP/TWALs.
+//
+// Considers two paths for inbound/outbound swaps.
+// 1) directly from tokenIn to tokenOut, or 2) from tokenIn to a hopToken to
+// tokenOut. hopTokens are other popular tokens that often appear in Uniswap
+// routes -- for example, WETH.
+//
+// This ensures that if tokenIn is a token with an active Uniswap pool with
+// either USDC or any of the hopTokens, the price is accurate.
+contract DaimoUSDCSwapper is IDaimoSwapper {
+  struct DaimoUSDCSwapperExtraData {
+    bytes swapPath;
+    uint128 altruisticAmountOut;
+    address altruisticSender;
+  }
+
+  // no-op function with struct as argument to expose it in generated ABI
+  function extraDataStruct(DaimoUSDCSwapperExtraData memory sig) public {}
+
+  uint256 constant _MAX_UINT128 = type(uint128).max;
+
+  // Constants used by Uniswap.
+  ISwapRouter public uniswapRouter;
+  IERC20 public usdc;
+  IERC20 public weth;
+  IERC20[] public hopTokens;
+  uint24[] public oracleFeeTiers;
+  uint32 public oraclePeriod;
+  IUniswapV3Factory public oraclePoolFactory;
+
+  event OracleError(address pool, uint32 secondsAgo, string reason);
+  event LowLevelOracleError(address pool, uint32 secondsAgo, bytes reason);
+
+  constructor(
+    IERC20 _usdc,
+    IERC20 _weth,
+    IERC20[] memory _hopTokens,
+    ISwapRouter _uniswapRouter,
+    uint24[] memory _oracleFeeTiers,
+    uint32 _oraclePeriod,
+    IUniswapV3Factory _oraclePoolFactory
+  ) {
+    usdc = _usdc;
+    weth = _weth;
+    hopTokens = _hopTokens;
+    uniswapRouter = _uniswapRouter;
+    oracleFeeTiers = _oracleFeeTiers;
+    oraclePeriod = _oraclePeriod;
+    oraclePoolFactory = _oraclePoolFactory;
+  }
+
+  // Gets TWAP and TWAL for a single pool.
+  function consultOracle(
+    address pool,
+    uint32 secondsAgo
+  )
+    public
+    view
+    returns (int24 arithmeticMeanTick, uint128 harmonicMeanLiquidity)
+  {
+    (arithmeticMeanTick, harmonicMeanLiquidity) = OracleLibrary.consult({
+      pool: pool,
+      secondsAgo: secondsAgo
+    });
+  }
+
+  // The best pool for a pair is the one with the highest liquidity over
+  // different fee tiers.
+  function getBestPoolTick(
+    IERC20 tokenA,
+    IERC20 tokenB
+  ) public view returns (address bestPool, int24 tick, uint24 bestFee) {
+    uint128 bestLiquidity = 0;
+    for (uint256 i = 0; i < oracleFeeTiers.length; i++) {
+      address pool = oraclePoolFactory.getPool({
+        tokenA: address(tokenA),
+        tokenB: address(tokenB),
+        fee: oracleFeeTiers[i]
+      });
```

Comment:
how do we handle symmetry? if there's a DAI-WETH pool and we're looking up (tokenA)WETH-(tokenB)DAI, will we still find it?

Reply:
Ah good question --

Reply:
i think this will still cause issues.
* `getPool` accepts tokens in either order
* however `getPool(WETH, USDT)` and `getPool(USDT, WETH)` will return the same pool address
* as a result, both of those will end up with same value for `arithmeticMeanTick`!
i think this means one of those two cases will end up with an incorrect output amount

Reply:
nvm, looks good! `getQuoteAtTick` handles multiplying or dividing by the tick price depending on `address(tokenA) < address(tokenB)`